### ISSUE TITLE
Compiler ignores preprocessor directives surrounding MSBuild directives in file-based apps

### ISSUE THREAD

**Issue by ericwj** [dotnet/roslyn#85203]
A classic case of "is this a bug, or is this a feature"?

I vote that this is a bug.

Seeing that the compiler wouldn't be invoked before the project file is available, debatable whether the minimal repro should generate an MSBuild warning, or work as written.

**Version Used**: 
```
error CS8304: Compiler version: '5.9.0-1.26379.115 (14fbf8d5271c98133561eb55185fdb05b286f578)'. Language version: 14.0.
```

**Steps to Reproduce**:

1. Compile the minimal repro below
2. Observe that it compiles and runs

A minimal repro, with source-code provided, is ideal.  Most compiler/language issues can be distilled into a snippet that can be pasted into [.NET Lab](https://lab.razor.fyi/#csharp).

```CS
#if false
#:package System.Threading.RateLimiting@10.0.11
#endif
using System.Threading.RateLimiting;
Console.WriteLine("Hello World!");
```

**Diagnostic Id**:

If this is a report about a bug in an analyzer, please include the diagnostic ID and message if possible (e.g. `"IDE0030: Use coalesce expression"`).

**Expected Behavior**:

```CS
#if false
//#:package System.Threading.RateLimiting@10.0.11
#endif
using System.Threading.RateLimiting;
Console.WriteLine("Hello World!");
```

```PS
PS C:\> dotnet csciffalse.cs
C:\csciffalse.cs(4,24): error CS0234: The type or namespace name 'RateLimiting' does not exist in the namespace 'System.Threading' (are you missing an assembly reference?)
```

**Actual Behavior**:

```PS
PS C:\> dotnet csciffalse.cs
Hello World!

PS C:\>
```

**Comment by jjonescz** [comment]
Thanks for reporting this.

Per the spec https://github.com/dotnet/csharplang/blob/main/proposals/csharp-14.0/ignored-directives.md, this should be an error:

> Ignored directives must also occur before any `#if` directives because the tooling might not know the full set of conditional compilation symbols while parsing ignored directives.

It looks like it indeed was originally an error but that regressed with https://github.com/dotnet/roslyn/pull/78158.

**Comment by AlekseyTs** [comment]
I do not think compiler should report any errors for inactive branches. The tooling, however, should follow the spec and should not attempt to   process the ignored directives that aren't at the beginning of the file. For the purpose of the feature, once `#if` is reached, that ends the "beginning" of the file. If that isn't clear from the spec, then the spec should be clarified.
I assume the tooling doesn't "see" ignored directive in the following code:
```
            System.Console.WriteLine("""
                #:package System.Threading.RateLimiting@10.0.11
                """);
```

Then it shouldn't "see" it in the repro code as well

**Comment by AlekseyTs** [comment]
Basically, I think that any directive that isn't an "ignored directive" should stop search for "ignored directives".

**Comment by AlekseyTs** [comment]
Related question. Is tooling supposed to understand comments? Especially multi-line comments?

**Comment by jjonescz** [comment]
> Related question. Is tooling supposed to understand comments? Especially multi-line comments?

What tooling? If you mean `dotnet run file.cs` tooling, then no. Not sure how it's related though.

**Comment by jjonescz** [comment]
Thanks @AlekseyTs, I think it's a good solution to not report any errors inside the disabled regions by roslyn. I have opened a PR in sdk to ensure we are not processing the directives in disabled regions (which previously only implicitly relied on roslyn diagnostics).

**Comment by AlekseyTs** [comment]
> What tooling? If you mean `dotnet run file.cs` tooling, then no. Not sure how it's related though.

It is somewhat related because comments also represent a portion of the document that should be ignored. For example, what is expected with this code?
```
#:package System.Something /*
#:package System.Threading.RateLimiting@10.0.11 */

using System.Threading.RateLimiting;
Console.WriteLine("Hello World!");
```

**Comment by jjonescz** [comment]
I see. The file-based app tooling is using roslyn lexer (via public API), so it definitely won't see that directive inside the comment. Only in disabled regions, roslyn lexer parses directives.
